java-checks/src/main/java/org/sonar/java/checks/security/AndroidPersistentUniqueIdentifierCheck.java

```diff
@@ -0,0 +1,125 @@
+import org.sonar.plugins.java.api.tree.MethodReferenceTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S7435")
+public class AndroidPersistentUniqueIdentifierCheck extends AbstractMethodDetection {
+
+  private static final String HARDWARE_ID_MESSAGE = "Using a hardware identifier puts user privacy at risk. Make sure it is safe here.";
+  private static final String PHONE_NUMBER_MESSAGE = "Using a phone number puts user privacy at risk. Make sure it is safe here.";
+  private static final String ADVERTISING_ID_MESSAGE = "Using Advertising ID puts user privacy at risk. Make sure it is safe here.";
+  private static final String NON_RESETTABLE_PERSISTENT_ID_MESSAGE = "Using a non-resettable persistent identifier puts user privacy at risk. Make sure it is safe here.";
+
+  private static final MethodMatchers staticSettingsSecureGetStringFunMatcher =
+    MethodMatchers.create()
+      .ofTypes("android.provider.Settings$Secure")
+      .names("getString")
+      .addParametersMatcher("android.content.ContentResolver", "java.lang.String")
+      .build();
+
+  private static final Map<MethodMatchers, String> matchers = Map.of(
+    MethodMatchers.create()
+      .ofTypes("android.bluetooth.BluetoothAdapter")
+      .names("getAddress")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.net.wifi.WifiInfo")
+      .names("getMacAddress")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.TelephonyManager")
+      .names("getSimSerialNumber", "getDeviceId", "getImei", "getMeid")
+      .withAnyParameters()
+      .build(),
+    HARDWARE_ID_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.TelephonyManager")
+      .names("getLine1Number")
+      .withAnyParameters()
+      .build(),
+    PHONE_NUMBER_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes("android.telephony.SubscriptionManager")
+      .names("getPhoneNumber")
+      .withAnyParameters()
+      .build(),
+    PHONE_NUMBER_MESSAGE,
+    MethodMatchers.create()
+      .ofSubTypes(
+        "com.google.android.gms.ads.identifier.AdvertisingIdClient$Info",
+        "androidx.ads.identifier.AdvertisingIdInfo",
+        "com.huawei.hms.ads.identifier.AdvertisingIdClient$Info"
+      )
+      .names("getId")
+      .withAnyParameters()
+      .build(),
+    ADVERTISING_ID_MESSAGE,
+    staticSettingsSecureGetStringFunMatcher,
+    NON_RESETTABLE_PERSISTENT_ID_MESSAGE
+  );
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.METHOD_INVOCATION, Tree.Kind.METHOD_REFERENCE);
+  }
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.or(matchers.keySet().toArray(MethodMatchers[]::new));
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    matchers.keySet().stream()
```

Comment:
[Same remark](https://github.com/SonarSource/sonar-kotlin/pull/600#discussion_r2012490830) here as for the corresponding PR in `sonar-kotlin`.
Using a map is OK here because matchers don't intersect, so there is no risk of matching a different matcher the day Map will change implementation and return a different order of keys.